gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java

```diff
@@ -267,4 +204,120 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
+  private scala.collection.Iterator<Row> getExpiredContentRows(
+      scala.collection.Iterator<String> references,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
+    List<Iterator<Row>> iterators = new ArrayList<>();
+    references.foreach(
+        reference -> {
+          iterators.add(computeExpiredContents(liveContentsBloomFilterMap, reference));
+          return reference;
+        });
+    // merge the iterators form each task.
+    Iterator<Row> mergedIterator = Iterators.concat(iterators.iterator());
+    return scala.collection.JavaConverters.asScalaIterator(mergedIterator);
+  }
+
+  private Iterator<Row> computeExpiredContents(
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap, String reference) {
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
+      return walkAllCommitsInReference(
+          api, GCUtil.deserializeReference(reference), liveContentsBloomFilterMap);
+    }
+  }
+
+  private Iterator<Row> walkAllCommitsInReference(
+      NessieApiV1 api,
+      Reference reference,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
+    Instant commitProtectionTime = Instant.now().minus(gcParams.getCommitProtectionDuration());
+    // Between the bloom filter creation and this step,
+    // there can be some more commits in the backend.
+    // Checking them against bloom filter will give false results.
+    // Hence, protect those commits using commitProtectionTime.
+    Predicate<LogResponse.LogEntry> unprotectedCommitsPredicate =
+        logEntry -> logEntry.getCommitMeta().getCommitTime().compareTo(commitProtectionTime) < 0;
+    // when no live bloom filter exist for this content id, all the contents are
+    // definitely expired.
+
+    // Checking against live contents bloom filter can result
+    //    no  --> content is considered expired (filter cannot say 'no' for live
+    // contents).
+    //    may be  -->  content is considered live (filter can say 'maybe' for expired
+    // contents).
+    // Worst case few expired contents will be considered live due to bloom filter
+    // fpp.
+    // But live contents never be considered as expired.
+    Predicate<Content> expiredContentPredicate =
+        content ->
+            (liveContentsBloomFilterMap.get(content.getId()) == null
+                || liveContentsBloomFilterMap.get(content.getId()).mightContain(content));
+    try {
+      Iterator<Content> iterator =
+          StreamingUtil.getCommitLogStream(
+                  api,
+                  builder ->
+                      builder
+                          .hashOnRef(reference.getHash())
+                          .refName(Detached.REF_NAME)
+                          .fetch(FetchOption.ALL),
+                  OptionalInt.empty())
+              .filter(unprotectedCommitsPredicate)
+              .map(LogResponse.LogEntry::getOperations)
+              .filter(operation -> operation instanceof Operation.Put)
```

Comment:
This is still wrong.

Reply:
Then how to filter the `PUT` operation ? can you please suggest? 
Do I need to implement a `operation.type` ?

Reply:
or you meant use this ? .filter(Operation.Put.class::isInstance)

Reply:
pushed changes now by using `Operation.Put.class::isInstance`

Reply:
Oh, it is a list of operations, so checking it against instance is wrong. I get it now !
Also figured out that test case didn't catch because validation is wrong.

It should be `dfExpected.except(dfActual)` instead of `dfActual.except(dfExpected)` because if the actual is empty, except() can still return empty.

Reply:
> It should be dfExpected.except(dfActual) instead of dfActual.except(dfExpected) because if the actual is empty, except() can still return empty.

Where's that??

Reply:
oh, In my code it is here:
`AbstractRestGC#verify()`